File: 23. Merge k Sorted Lists/solution.py
# Definition for singly-linked list.
class ListNode(object):

    def __init__(self, x):
        self.val = x
        self.next = None


class Solution(object):

    def mergeKLists(self, lists):
        """
        :type lists: List[ListNode]
        :rtype: ListNode
        """
        total = []
        for head in lists:
            total += self.listToArray(head)

        total.sort()

        return self.arrayToList(total)

    def getCurrentMinNode(self, lists):
        tmp = [(i, a) for i, a in enumerate(lists) if a is not None]
        if not tmp:
            return None
        minIndex, minNode = tuple(min(tmp, key=lambda elem: elem[1].val))
        lists[minIndex] = lists[minIndex].next
        return minNode

    def listToArray(self, head):
        tmp = []
        ptr = head
        while ptr != None:
            tmp.append(ptr.val)
            ptr = ptr.next
        return tmp

    def arrayToList(self, array):
        if not array:
            return None
        for i, v in enumerate(array):
            if i == 0:
                head = ListNode(v)
                pointer = head
            else:
                pointer.next = ListNode(v)
                pointer = pointer.next

        return head


def printList(head):
    pointer = head
    while pointer.next is not None:
        print pointer.val, "->",
        pointer = pointer.next
    print pointer.val


if __name__ == '__main__':
    so = Solution()
    test = [[1]]
    testlists = [so.arrayToList(i) for i in test]
    head = so.mergeKLists(testlists)
    printList(head)

File: 23. Merge k Sorted Lists/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
private:
    struct {
        bool operator()(const ListNode* a, const ListNode *b) {
            return a->val > b->val;
        }
    } cmp;

public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        ListNode dummy(-1);
        ListNode *cur = &dummy;
        priority_queue<ListNode*, vector<ListNode*>, decltype(cmp)> pq(cmp);

        for (auto &l: lists) {
            if (l) pq.push(l);
        }
        if (pq.empty()) return nullptr;
        while (!pq.empty()) {
            cur = cur->next = pq.top();
            pq.pop();
            if (cur->next) pq.push(cur->next);
        }
        return dummy.next;
    }
};

class Soltuion {
public:
    ListNode* mergeKLists(vector<ListNode*> &lists) {
        ListNode dummy(-1);
        ListNode *cur = &dummy;
        for (auto &l: lists) {
            cur->next = l;
            while (cur->next != nullptr)
                cur = cur->next;
        }
        std::sort(dummy.next, [](const ListNode *a, const ListNode *b) -> bool {
            return a->val < b->val;
        });
        return res;
    }
};